extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java

```diff
@@ -88,6 +91,11 @@
   private final AWSCredentialsConfig awsCredentialsConfig;
   private volatile Map<String, Long> currentPartitionTimeLag;
 
+  // Maintain sets of currently closed shards to find "bad" (closed and empty) shards
+  // Poll closed shards once and store the result to avoid redundant costly calls to kinesis
+  private final Set<String> emptyClosedShardIds = new TreeSet<>();
```

Comment:
Should these be in thread-safe container, or have access be protected with lock?

Reply:
Thanks for pointing this out, @zachjsh.
I think a ConcurrentHashSet would suffice here.

cc: @AmatyaAvadhanula

Reply:
Synchronizing the whole method `updateClosedShardCache` would actually be preferable because the state returned by two subsequent calls to `recordSupplier.getShards()` can be different.
So this call should happen inside the synchronized block, as should the calls to `recordSupplier.isClosedShardEmpty()`.

I hope this doesn't cause bottlenecks though.

Reply:
The whole method has been synchronized. Thanks!